### PR TITLE
Upgrade Byte Buddy to 1.9.0

### PR DESCRIPTION
**Pull request by tylerbenson**
Includes some bug fixes and performance improvements.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -182,7 +182,6 @@ jobs:
           # Reset the cache approx every release
           keys:
             - dd-trace-java-version-scan-{{ checksum "dd-trace-java.gradle" }}
-            - dd-trace-java-version-scan
```
Comment (realark) [inline .circleci/config.yml]: We were caching version scan gradle caches before and we're not now? I don't follow.

**Comment by tylerbenson** [comment]
@raphw,  I'm trying to upgrade to the latest byte buddy, but it seems the `OuterType` validation is breaking the tests in this new version.  Getting the following exception:

```
java.lang.IllegalStateException: size = 0
	at net.bytebuddy.matcher.FilterableList$AbstractBase.getOnly(FilterableList.java:108)
	at net.bytebuddy.pool.TypePool$Default$LazyTypeDescription$TypeContainment$WithinMethod.getEnclosingMethod(TypePool.java:3025)
	at net.bytebuddy.pool.TypePool$Default$LazyTypeDescription.getEnclosingMethod(TypePool.java:2727)
	at net.bytebuddy.description.type.TypeDescription$AbstractBase$OfSimpleType$WithDelegation.getEnclosingMethod(TypeDescription.java:7142)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining$WithFullProcessing$RedefinitionClassVisitor.onOuterType(TypeWriter.java:3972)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining$WithFullProcessing$RedefinitionClassVisitor.onVisitOuterClass(TypeWriter.java:3967)
	at net.bytebuddy.utility.visitor.MetadataAwareClassVisitor.visitOuterClass(MetadataAwareClassVisitor.java:104)
	at net.bytebuddy.jar.asm.ClassReader.accept(ClassReader.java:550)
	at net.bytebuddy.jar.asm.ClassReader.accept(ClassReader.java:391)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:3028)
	at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:1729)
	at net.bytebuddy.dynamic.scaffold.inline.RedefinitionDynamicTypeBuilder.make(RedefinitionDynamicTypeBuilder.java:190)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Transformation$Simple$Resolution.apply(AgentBuilder.java:9203)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:9614)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:9577)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1400(AgentBuilder.java:9345)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$LegacyVmDispatcher.run(AgentBuilder.java:9942)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$LegacyVmDispatcher.run(AgentBuilder.java:9891)
	at java.security.AccessController.doPrivileged(Native Method)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:9500)
	at sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:428)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:338)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:411)
	at datadog.trace.agent.test.SpockRunner$InstrumentationClassLoader.loadClass(SpockRunner.java:187)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at JaxRsAnnotationsInstrumentationTest.$spock_feature_1_0prov1(JaxRsAnnotationsInstrumentationTest.groovy:49)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:188)
	at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:84)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481)
	at org.spockframework.runtime.ParameterizedSpecRunner.createDataProviders(ParameterizedSpecRunner.java:57)
	at org.spockframework.runtime.ParameterizedSpecRunner.runParameterizedFeature(ParameterizedSpecRunner.java:38)
	at org.spockframework.runtime.BaseSpecRunner.doRunFeature(BaseSpecRunner.java:262)
	at org.spockframework.runtime.BaseSpecRunner$5.invoke(BaseSpecRunner.java:246)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:464)
	at org.spockframework.runtime.BaseSpecRunner.runFeature(BaseSpecRunner.java:238)
	at org.spockframework.runtime.BaseSpecRunner.runFeatures(BaseSpecRunner.java:188)
	at org.spockframework.runtime.BaseSpecRunner.doRunSpec(BaseSpecRunner.java:98)
	at org.spockframework.runtime.BaseSpecRunner$1.invoke(BaseSpecRunner.java:84)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:481)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:464)
	at org.spockframework.runtime.BaseSpecRunner.runSpec(BaseSpecRunner.java:76)
	at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:67)
	at org.spockframework.runtime.Sputnik.run(Sputnik.java:63)
	at datadog.trace.agent.test.SpockRunner.run(SpockRunner.java:92)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

It specifically doesn't seem to like the anonymous classes that are defined inline of our spock tests:
https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/jax-rs-annotations/src/test/groovy/JaxRsAnnotationsInstrumentationTest.groovy

Any suggestions for ways around this, or do I need to pull the classes out of the spock test?

Thanks for your help.

File: .circleci/config.yml (hunk 1 above)
Comment (tylerbenson) [reply .circleci/config.yml]: The cache for version scan/muzzle is getting pretty big.  This allows the cache to fully reset after each release.  Otherwise it always pulls in the cache from the previous release and continues building on it.

**Comment by raphw** [comment]
Seems like Groovy creates class files that are declared as local types within a method but where the method is not declared. The call to `getEnclosingMethod` indicates that the class file expects a method being declared by some class that does not exist in this type.

Could you extract the class file of such a class and post it to this issue? Also, if you identified such a class you you try and see what happens on calling: `Class.forName("...").getEnclosingMethod()`? I would guess that the reflection API also throws an exception.

**Comment by tylerbenson** [comment]
@raphw Here's the whole directory zipped up.  It contains a bunch of those class files.
[test.zip](https://github.com/DataDog/dd-trace-java/files/2359576/test.zip)

I checked the `getEnclosingMethod` thing and on each of [these](https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/jax-rs-annotations/src/test/groovy/JaxRsAnnotationsInstrumentationTest.groovy#L49-L84) I get a `java.lang.InternalError: Enclosing method not found`.  The last test case however returns null since it's not an anonymous class.

Thanks for your help!

**Comment by raphw** [comment]
There seems to be a serious bug in the Groovy compiler. For example `JaxRsAnnotationsInstrumentationTest` defines a method `$spock_feature_1_1(java.lang.Object)` that has an annotation with the value *span named '#name' from annotations on class*. Within this method, the anonymous class `JaxRsAnnotationsInstrumentationTest$1` which defines an enclosing method with the annotation value and a signature of `()Ljava/lang/Object;`.

This method does of course not exist and both the JVM and Byte Buddy do not know how to handle this error and throw an exception or error respectively.

I will try to come up with a workaround in Byte Buddy. I think the easiest solution would be to catch any throwable on that call site and to retain the original information of the class file if reading the property is not possible. I do not love this solution as this can also yield false positives but I do actually now remember an issue that is probably related from Mockito. This broken mechanism seems to be spread.

While I fix this issue in Byte Buddy, could you spend a few minutes and report it to Groovy. That their code breaks the reflection API should make up for a good reason to fix this even without Byte Buddy involved.

**Comment by tylerbenson** [comment]
@raphw thanks for looking into this!

Before I submit a bug report, is there any chance this could be caused by Spock and not groovy? (I’m not sure how Spock does it’s magic.)

**Comment by raphw** [comment]
I'm fairly sure its Groovy, by using the AST API you cannot declare class properties like this. I think they mix up the Groovy name and the class file name of the method.

Pushing out a local fix tonight.

**Comment by raphw** [comment]
Releasing 1.8.22 right now which should work around this problem.

**Comment by tylerbenson** [comment]
@raphw awesome thanks. I’ll try it out this week.

**Comment by raphw** [comment]
Interesting, it seems like Scala defines an anonymous class that is also a top-level class. From looking into the code, this is the only explanation for this error.

Could you send me a copy of the class file that causes the error? Either attached to this ticket or to rafael(dot)wth(at)gmail(dot)com.

You might also be interested in trying the [`DECORATE`](https://github.com/raphw/byte-buddy/blob/master/byte-buddy-dep/src/main/java/net/bytebuddy/agent/builder/AgentBuilder.java#L2043) strategy for transforming for your use case if you do not change the class layout. I'd be still thankful, though, if you could help me fix this release. The inner type attribute is really tricky since it is not enforced by the verifier such that many non-Java languages do not really get it right.

**Comment by raphw** [comment]
I think I managed to reproduce this and implemented a fix. Wil be a part of Byte Buddy 1.9.0 which will be released next week.

**Comment by tylerbenson** [comment]
@raphw sorry for the delay.  I was traveling all weekend. Here's the class file: 
[AsyncExecutor$anon$2$anon$3.class.zip](https://github.com/DataDog/dd-trace-java/files/2411962/AsyncExecutor.anon.2.anon.3.class.zip)

If you have a snapshot published somewhere, I'd be happy to try it out on this branch. We seem to have pretty good edge case test coverage.  😉 

Let me know if you need anything else.

**Comment by raphw** [comment]
It is strange that enclosing type is null as the attribute is set. I assume that there is a particularity on how the class file is processed, the attribute combination is a bit strange.

Can you try with the current master? Its already set for 1.9.0, there will be some API changes but you will probably need to incorporate those anyways. I wonder if the previous fix also took care of this issue.

**Comment by tylerbenson** [comment]
@raphw do you publish snapshots anywhere on merge to master? That would make it much easier to try out.

**Comment by raphw** [comment]
I find it easier to use JitPack: https://jitpack.io/#raphw/byte-buddy

**Comment by tylerbenson** [comment]
@raphw it seems the master build passed our tests after the required API change updates.  Nice work!  What's the ETA on the release?

**Comment by raphw** [comment]
ASM 7 was released today and I will release tonight.

**Comment by tylerbenson** [comment]
@raphw awesome, thanks!

**Comment by realark** [comment]
@tylerbenson We should probably update our version of ASM as well.
